### PR TITLE
add precommit

### PR DESCRIPTION
**Pull request by CoderZhi**
Add an interface PreCommitter. Protocols which are PreCommitter will be called before a workingset commit.

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/protocol/staking/candidate_center.go [hunk 1]
```diff
@@ -435,7 +439,27 @@ func (cb *candBase) all() CandidateList {
...
+		clone.operatorMap[operator] = cand.Clone()
+	}
+	for bucket, cand := range cb.selfStkBucketMap {
+		clone.selfStkBucketMap[bucket] = cand.Clone()
+	}
+
+	return clone
+}
+
+func (cb *candBase) Commit(change *candChange) (int, error) {
```
Comment (dustinxie) [inline action/protocol/staking/candidate_center.go:462]: all funcs of `candBase` are private now, let's use `commit()` to be consistent?

File: action/protocol/staking/candidate_center.go [hunk 2]
```diff
@@ -435,7 +439,27 @@ func (cb *candBase) all() CandidateList {
 	return list
 }
 
-func (cb *candBase) commit(change *candChange) (int, error) {
+func (cb *candBase) Clone() *candBase {
```
Comment (dustinxie) [inline action/protocol/staking/candidate_center.go:442]: same here, use `clone`

File: state/factory/workingset.go [hunk 3]
```diff
@@ -216,10 +216,14 @@ func (ws *workingSet) ResetSnapshots() {
...
 // Commit persists all changes in RunActions() into the DB
 func (ws *workingSet) Commit(ctx context.Context) error {
+	if err := protocolPreCommit(ctx, ws); err != nil {
+		return err
+	}
 	if err := ws.store.Commit(); err != nil {
 		return err
 	}
 	if err := protocolCommit(ctx, ws); err != nil {
+		// TODO (zhi): wrap the error and eventually panic it in caller side
```
Comment (dustinxie) [inline state/factory/workingset.go:226]: could you pls file an issue? can be fixed in next hard-fork

File: action/protocol/staking/candidate_center.go (hunk 1 above)
Comment (CoderZhi) [reply action/protocol/staking/candidate_center.go:462]: do not understand why a private struct needs to have private functions

**Review by raullenchai — APPROVED** [review]

**Comment by dustinxie** [comment]
this code has been integrated into (#3661), so close it

**Pull request closed without merging**